# Cecret

Named after the beautiful [Cecret lake](https://en.wikipedia.org/wiki/Cecret_Lake) 

Location: 40.570°N 111.622°W , 9,875 feet (3,010 m) elevation


Cecret is UPHL's workflow for SARS-COV-2 sequencing with the artic/Illumina hybrid library prep workflow for MiSeq data. Built to work on linux-based operating systems. 

# Getting started

```
git clone https://github.com/UPHL-BioNGS/Cecret.git
```

To make your life easier, follow with

```
cd Cecret
git init
```

So that you can use `git pull` can be used for updates.

## Prior to starting the workflow

### install dependencies
- [Nextflow](https://www.nextflow.io/docs/latest/getstarted.html) 
- [Singularity](https://singularity.lbl.gov/install-linux)

### create covid_samples.txt

This pipeline is meant to enable quick and easy file renaming for sample submission to public repositories. As such, a text file with the following columns (no header please): `Laboratory_id submission_id collection_date`

Example covid_samples.txt file contents:
```
123456  UT-NA001  2020-06-10
```
Where the files named `123456-UT-M03999-200602_S9_L001_R1_001.fastq.gz`, `123456-UT-M03999-200602_S9_L001_R2_001.fastq.gz` will be renamed `UT-NA001.fastq.R1.fastq.gz` and `UT-NA001.fastq.R2.fastq.gz` with the corresponding consensus file named `UT-NA001.consensus.fa`. Collection date is used in Genbank fasta file creation.

### arrange the fastq.gz reads
```
directory
|-covid_samples.txt
|-Sequencing_reads
   |-Raw
     |-*fastq.gz
```

# Usage
```
nextflow run Cecret.nf -c config/cecret.singularity.nextflow.config
```

The main components of Cecret are:

- [seqyclean](https://github.com/ibest/seqyclean) - for cleaning reads
- [bwa](http://bio-bwa.sourceforge.net/) - for aligning reads to the reference
- [ivar](https://andersen-lab.github.io/ivar/html/manualpage.html) - for trimming primers, calling variants, and creating a consensus fasta
- [samtools](http://www.htslib.org/) - for QC metrics and sorting
- [fastqc](https://github.com/s-andrews/FastQC) - for QC metrics
- [bedtools](https://bedtools.readthedocs.io/en/latest/) - for depth estimation over amplicons

## Final file structure
```
directory
|-covid_samples.txt
|-run_results.txt                       # information about the sequencing run that's compatible with legacy workflows
|-fastqc
| |-sample.fastqc.html
| |-sample.fastqc.zip
|-logs
| |-process_logs                        # for troubleshooting puroses
|-Sequencing_reads
| |-Raw
| | |-sample_S1_L001_R1_001.fastq.gz    # initial file
| | |-sample_S1_L001_R2_001.fastq.gz    # inital file
| |-QCed
|   |-sample_clean_PE1.fastq            # clean file
|   |-sample_clean_PE2.fastq            # clean file
|-covid
  |-bedtools
  | |-multicov.txt                      # coverage information for each amplicon
  |-bwa
  | |-sample.sorted.bam                 # aligned reads
  | |-sample.sorted.bam.bai             # indexes
  |-consensus
  | |-sample.consensus.fa               # consensus fasta file generated by ivar
  |-samtools_coverage
  | |-bwa
  | | |-sample.cov.hist                 # histogram of coverage for aligned reads
  | | |-sample.cov.txt                  # tabular information of coverage for aligned reads
  | |-sort
  |   |-sample.cov.trim.hist            # histogram of coverage for aligned reads after primer trimming
  |   |-sample.cov.trim.txt             # tabular information of coverage for aligned reads after primer trimming
  |-samtools_stats
  | |-bwa
  | | |-sample.stats.txt                # samtools stats for aligned reads
  | |-sort
  |   |-sample.stats.trim.txt           # samtools stats for trimmed reads
  |-sorted
  | |-sample.primertrim.sorted.bam      # aligned reads after primer trimming and sorting
  |-submission_files
  | |-run_id.genbank_submission.fasta   # multifasta for direct genbank
  | |-run_id.gisaid_submission.fasta    # multifasta file bulk upload to gisaid
  | |-sample.consensus.fa               # renamed consensus fasta file
  | |-sample.genbank.fa                 # fasta file with formatting and header including metadata for genbank
  | |-sample.gisaid.fa                  # fasta file with header for gisaid
  | |-sample.R1.fastq.gz                # renamed raw fastq.gz file
  | |-sample.R2.fastq.gz                # renamed raw fastq.gz file
  |-summary
  | |-sample.run_results.txt
  | |-sample.summary.txt
  |-trimmed
  | |-sample.primertrim.bam             # aligned reads after primer trimming
  |-variants
  | |-sample.variants.tsv               # list of variants identified via ivar and corresponding scores
  |-summary.txt                         # comma separated file with coverage, depth, number of failed amplicons, and number of "N" information
```

# Hybrid run instructions

Sometimes SARS-COV-2 sequences are run with other samples on the MiSeq. When [UPHL's reference free pipeline]() is run prior to this analysis, instead use the following command:
```
nextflow run Cecret_partial.nf -c config/partial.singularity.nextflow.config
```

This partial workflow uses [blobtools](https://blobtools.readme.io/docs) in addition to the above tools for identification of reads.

# Adjustable Paramters

All workflow parameters can be adjusted in a config file or on the command line as follows:

```
nextflow run Cecret.nf -c config/cecret.singularity.nextflow.config --outdir new_directory
```

Adjustable workflow paramaters with default values:
```
artic_version = 'V3'
year = '2020'
outdir = current_directory
log_directory = current_directory + '/logs'
sample_file = current_directory/covid_samples.txt
```
For use with staphb/seqyclean:1.10.09 container run with singularity
```
seqyclean_contaminant_file="/Adapters_plus_PhiX_174.fasta"
```
For use with staphb/ivar:1.2.2_artic20200528 container run with singularity
```
primer_bed = "/artic-ncov2019/primer_schemes/nCoV-2019/${params.artic_version}/nCoV-2019.bed"
reference_genome = "/artic-ncov2019/primer_schemes/nCoV-2019/${params.artic_version}/nCoV-2019.reference.fasta"
gff_file = "/reference/GCF_009858895.2_ASM985889v3_genomic.gff"
amplicon_bed = "/artic-ncov2019/primer_schemes/nCoV-2019/${params.artic_version}/nCoV-2019_amplicon.bed"
```

### Other useful options
To create a report, use `-with-report` with your nextflow command.

# References

![alt text](https://uphl.utah.gov/wp-content/uploads/New-UPHL-Logo.png)
